### ISSUE TITLE
Formally specify Sass and SCSS

### ISSUE THREAD

**Issue by yukulele** [sass/sass#765]
_Edited by @nex3 to change the title._

Hi,
I think SASS should be a specification that developers are free to implement in any programming language.

Ruby SASS compiler should be one of his implementations.

This may help other implementations (PHPSass, sass.js, pyScss, ...) to stay up to date.

some web designer are interested by SASS but can't use it because they don't work with Ruby...
Current other implementations are sometime buggy and rarely updated...

**Comment by chriseppstein** [comment]
:thumbsup: But it's a lot of work. Maybe as part of 4.0 release.

**Comment by nex3** [comment]
I'm skeptical that this will happen at all. Writing a formal specification is a colossal amount of work. That work has to be done by someone -- presumably @chriseppstein or myself -- and it would come straight out of time that could otherwise be spent improving the language. The value provided by a specification is just not large enough to warrant the effort.

It's also worth noting that a specification won't make it easier to port Sass to other languages; it will just make it easier for those ports to be confident that they're handling every edge case correctly. Currently I don't know of any Sass ports that have feature parity with the reference implementation. Without ports that can make use of it, writing a spec provides no value at all.

**Comment by chriseppstein** [comment]
Libsass is very close to feature parity. I want to provide a single set of test cases that works for sass and libsass and anything else that wants to write an adapter. It doesn't have to be a formal spec. But it would be good to know whether some implementation is roughly equivalent.

**Comment by nex3** [comment]
I have no problem with a set of common test cases, but that's very different from a specification.

**Comment by chriseppstein** [comment]
True, I was thinking of spec in the way that is used in the project [rubyspec](http://rubyspec.org/).

**Comment by yukulele** [comment]
First, it's a versioning problem, each new SASS version add new languages features and ruby specific feature (code optimisation, bug corrections, ...) the both must to be distinctly separated

For exemple in <a href="http://sass-lang.com/docs/yardoc/file.SASS_CHANGELOG.html#329">the last version</a>, only one line is a SASS specification (for compiling) others are ruby specific bug fix:

```
v. 3.2.9

(ruby) Fix a bug where @extends would occasionally cause a selector to be generated with the incorrect specificity.
(ruby) Avoid loading listen v1.0, even if it’s installed as a Gem (see issue 719).
(ruby) Update the bundled version of listen to 0.7.3.
(sass) Automatically avoid the IE7 content: counter bug.
```

Symply specify type of new features in versions (like <a href="http://www.mozilla.org/en-US/firefox/21.0/releasenotes/">firefox</a>) whould be the first (and big) step toward language specification.

**Comment by nex3** [comment]
It's not especially difficult to read the changelog and determine which changes are language-based and which are implementation-based.

**Comment by richthegeek** [comment]
I'm passing the buck a little here, but in response to "Currently I don't know of any Sass ports that have feature parity with the reference implementation", part of the problem with keeping a port up to date is not knowing precisely what has changed over the course of several versions.

To solve this requires either specification which is in some standard notation (for example, that used for any context-free grammar: http://en.wikipedia.org/wiki/Context-free_grammar), or a COMPLETE test specification.

Without it, any hope for a parallel implementation (and widespread use of SASS outside of the ruby sphere) is a little difficult to envisage.

**Comment by nex3** [comment]
@richthegeek Formal specifications or comprehensive cross-platform test suites are useful for ensuring that every relevant edge case works identically between implementations. They're not necessary for getting basic feature parity, which no ports have yet reached. Until they have, work spent on a specification isn't providing value to anyone.

It's also worth noting that the Sass syntax is not context-free, so it can't be represented entirely in BNF.

**Comment by techdragon** [comment]
A comprehensive low level test suite built in a way that by design does not depend on any particular sass implementation would be best for helping other projects benefit quickly from the effort. Ideally the tests could be run anywhere a sass compiler can be run. This may be the easiest way to accomplish the task as it could be simplified so that it limits the need to rework things beyond the test cases already in use by ruby sass, which would avoid it being a maintenance burden since the test cases could transition from text inside ruby files into standalone files that ruby opened & used, then theres no waste & the language spec tests can be both used by ruby sass and reused by others.  

We could get the test runner down to something as portable as this. Just test case scss files, expected output css files, and a simple test runner kinda like the one below. 

``` bash
#!/bin/bash
$1=SASS_COMPILER

TESTS_RUN=0
TESTS_PASSED=0
TESTS_FAILED=0

TEST_CASES=(./sass-test-cases/input/*.scss)
NUMBER_OF_TESTS=${#TEST_CASES[@]}

RED_TEXT='\e[0;31m'
GREEN_TEXT='\e[0;32m'
CLEAR_COLOURS='\e[0m'

for TEST_CASE_SASS_FILE in $TEST_CASES
do
    TEST_CASE=$(basename ${TEST_CASE_SASS_FILE} .scss)

    $SASS_COMPILER ./sass-test-cases/input/${TEST_CASE}.scss ./sass-test-output/${TEST_CASE}.css

    cmp -s ./sass-test-cases/ouput/${TEST_CASE}.css ./sass-test-output/${TEST_CASE}.css

    RETURN_CODE = $?

    if [ RETURN_CODE -eq 0 ]
    then
        RESULT="Passed"
        COLOUR=${GREEN_TEXT}
        TESTS_PASSED++
        TESTS_RUN++
    else
        RESULT="Failed"
        COLOUR=${RED_TEXT}
        TESTS_FAILED++
        TESTS_RUN++
    elif

    echo "${COLOUR}Test ${TESTS_RUN} of ${NUMBER_OF_TESTS} - ${RESULT} - ${TEST_CASE}.scss ${RESULT}${CLEAR_COLOURS}"

done
```

**Comment by yukulele** [comment]
https://github.com/sass/sass-spec

done?

**Comment by nex3** [comment]
@yukulele That's a test suite, which isn't the same as a formal prose specification.
